// @namra_ace/zen-auth/dist/adapters/MongoStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MongoStore = void 0;
class MongoStore {
    constructor(mongooseModel) {
        this.model = mongooseModel;
    }
    async findAllByUser(userId) {
        const docs = await this.model.findOne({ userId });
        if (!docs)
            return [];
        return docs.map((doc) => doc.data);
    }
    async deleteByUser(userId) {
        await this.model.deleteOne({ userId });
    }
    async set(key, value, ttlSeconds) {
        const expiresAt = new Date(Date.now() + ttlSeconds * 1000);
        // Upsert: Update if exists, Insert if new
        await this.model.updateOne({ _id: key }, { _id: key, data: value, expiresAt }, { upsert: true });
    }
    async get(key) {
        const doc = await this.model.findOne({ _id: key });
        if (!doc)
            return null;
        // MongoDB TTL indexes usually handle cleanup, but we double-check here
        if (new Date() > doc.expiresAt) {
            return null;
        }
        return doc.data;
    }
    async delete(key) {
        await this.model.deleteOne({ _id: key });
    }
    async touch(key, ttlSeconds) {
        const expiresAt = new Date(Date.now() + ttlSeconds * 1000);
        // The "Slide": Just update the date
        await this.model.updateOne({ _id: key }, { $set: { expiresAt } });
    }
}
exports.MongoStore = MongoStore;
